### PR TITLE
[feature](nerieds)Support partition prune

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: noissue

## Problem summary

Rewrite filter->project to project -> filter to make sure when do partition prune the tree looks like this: project -> filter -> olapscan.
Enable the `MergeConsecutiveProject` `MergeConsecutiveFilter`.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by wangshuo128** [comment]
Please add unit tests to cover tests in `ListPartitionPrunerTest` and `RangePartitionPruneTest`.

**Comment by Kikyou1997** [comment]
@morrySnow

**Comment by Kikyou1997** [comment]
> Please add unit tests to cover tests in `ListPartitionPrunerTest` and `RangePartitionPruneTest`.

Current support  RangePartitionPrune first in this PR

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 1]
```diff
@@ -72,6 +72,8 @@ public enum RuleType {
     MERGE_CONSECUTIVE_FILTERS(RuleTypeClass.REWRITE),
     MERGE_CONSECUTIVE_PROJECTS(RuleTypeClass.REWRITE),
     REWRITE_SENTINEL(RuleTypeClass.REWRITE),
+    PARTITION_PRUNE(RuleTypeClass.REWRITE),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:75]:
```suggestion
    OLAP_SCAN_PARTITION_PRUNE(RuleTypeClass.REWRITE),
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/RewriteAliasToChildExpr.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.apache.doris.nereids.trees.expressions.Alias;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+
+import java.util.Map;
+
+/**
+ * UpdateAliasVisitor
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/RewriteAliasToChildExpr.java:28]: detailing it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 3]
```diff
@@ -55,6 +55,7 @@ public LogicalOlapScan(Table table, List<String> qualifier) {
     public LogicalOlapScan(Table table, List<String> qualifier,
                            Optional<GroupExpression> groupExpression, Optional<LogicalProperties> logicalProperties) {
         super(PlanType.LOGICAL_OLAP_SCAN, table, qualifier, groupExpression, logicalProperties);
+        super.selectedPartitionIds = ((OlapTable) table).getPartitionIds();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:58]: why not add a new parameter to super constructor directly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java [hunk 4]
```diff
@@ -42,6 +44,8 @@ public abstract class LogicalRelation extends LogicalLeaf implements Scan {
     protected final Table table;
     protected final List<String> qualifier;
 
+    protected Collection<Long> selectedPartitionIds = Lists.newArrayList();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRelation.java:47]: why need collection?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java [hunk 5]
```diff
@@ -117,11 +121,15 @@ public <R, C> R accept(PlanVisitor<R, C> visitor, C context) {
...
+                selectedPartitionId, qualifier, distributionSpec, groupExpression, logicalProperties);
     }
 
     @Override
     public Plan withLogicalProperties(Optional<LogicalProperties> logicalProperties) {
-        return new PhysicalOlapScan(olapTable, qualifier, distributionSpec, Optional.empty(), logicalProperties.get());
+        return new PhysicalOlapScan(olapTable,
+                selectedPartitionId,
+                qualifier,
+                distributionSpec, Optional.empty(), logicalProperties.get());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalOlapScan.java:133]:
```suggestion
        return new PhysicalOlapScan(olapTable, selectedPartitionId, qualifier,
                distributionSpec, Optional.empty(), logicalProperties.get());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+/**
+ * Used to prune partition of olap scan, should execute after SwapProjectAndFilter, MergeConsecutiveFilters,
+ * MergeConsecutiveProjects and all predicate push down related rules.
+ */
+public class PruneOlapScanPartition extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalOlapScan()).thenApply(ctx -> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:65]: if u do not use other attributes in `ctx`, u could use `then()`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            for (String colName : partitionColumnNameSet) {
+                ColumnRange columnRange = createColumnRange(colName, expressionList);
+                columnNameToRange.put(colName, columnRange);
+            }
+
+            Map<Long, PartitionItem> keyItemMap = partitionInfo.getIdToItem(false);
+            PartitionPruner partitionPruner = new RangePartitionPrunerV2(keyItemMap,
+                    partitionInfo.getPartitionColumns(), columnNameToRange);
+            Collection<Long> selectedPartitionId =  Utils.execWithReturnVal(partitionPruner::prune);
+            scan.getSelectedPartitionIds().retainAll(selectedPartitionId);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:89]: REMEMBER THAT: `Plan` and `Expression` in Nereids is **_IMMUTABLE_**.
we should not mod scan's attribute. Instead, we need to generate a new `LogicalOlapScan` with pruned partitionIds.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            return ctx.root;
+        }).toRule(RuleType.PARTITION_PRUNE);
+    }
+
+    private ColumnRange createColumnRange(String colName, List<Expression> expressionList) {
+        ColumnRange result = ColumnRange.create();
+        for (Expression expression : expressionList) {
+            Expression leftMostChild = expression.leftMostNode();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:97]: use `collect(SlotReference.class:isInstance)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+ */
+public class PruneOlapScanPartition extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalOlapScan()).thenApply(ctx -> {
+            LogicalFilter<LogicalOlapScan> filter = ctx.root;
+            LogicalOlapScan scan = filter.child();
+            Expression predicate = filter.getPredicates();
+            List<Expression> expressionList = ExpressionUtils.extractConjunction(predicate);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:69]: move these after L79 for ease reading

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                    default:
+                        break;
+                }
+            }
+        }
+        return result;
+    }
+
+    private ColumnRanges exprToRanges(Expression expression, String colName) {
+        // TODO: process in/is null expression
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:137]: is these needed by TPC-H?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            Expression leftMostChild = expression.leftMostNode();
+            if (!(leftMostChild instanceof SlotReference)) {
+                continue;
+            }
+            SlotReference slotRef = (SlotReference) leftMostChild;
+            if (!slotRef.boundToColumn(colName)) {
+                continue;
+            }
+            if (expression instanceof Or) {
+                ColumnRanges ranges = exprToRanges(expression, colName);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:106]: this is strange that `expression` is a `Or`, but `exprToRanges` need a `ComparisonPredicate `.
furthermore, if and else branches are same.
So, do u forget something?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        LiteralExpr value = ((Literal) rightChild).toLegacyLiteral();
+        if (expression instanceof EqualTo) {
+            ColumnBound bound = ColumnBound.of(value);
+            result.add(Range.closed(bound, bound));
+        } else if (expression instanceof GreaterThanEqual) {
+            result.add(Range.atLeast(ColumnBound.of(value)));
+        } else if (expression instanceof GreaterThan) {
+            result.add(Range.greaterThan(ColumnBound.of(value)));
+        } else if (expression instanceof LessThan) {
+            result.add(Range.atMost(ColumnBound.of(value)));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:156]:
```suggestion
            result.add(Range.LessThan(ColumnBound.of(value)));
```

add more test please

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+ * Rewrite filter -> project to project -> filter.
+ */
+public class SwapFilterAndProject extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalProject()).thenApply(ctx -> {
+            LogicalFilter<LogicalProject<GroupPlan>> filter = ctx.root;
+            LogicalProject<GroupPlan> project = filter.child();
+            List<NamedExpression> namedExpressionList = project.getProjects();
+            Map<Slot, Alias> slotToAlias = new HashMap<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:47]:
```suggestion
            Map<Slot, Alias> slotToAlias = Maps.newHashMap();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+            namedExpressionList.stream().filter(Alias.class::isInstance).forEach(s -> {
+                slotToAlias.put(s.toSlot(), (Alias) s);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:49]: maybe u could generate a `Map<Expression, Expression>` like `(s.toSlot, s.child())`, and use `ExpressionReplacer.INSTANCE` to replace slot reference in expression

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java (hunk 10 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:137]: It's determined on how the the partition is set for tpch table.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java (hunk 13 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:47]: What's the difference between these two way?

Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:47]: nothing, just personal preference

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java (hunk 9 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:70]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java (hunk 8 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java:98]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/RewriteAliasToChildExpr.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+
+import java.util.Map;
+
+/**
+ * Rewrite out slot of alias to alias's child expr.
+ */
+public class RewriteAliasToChildExpr extends DefaultExpressionRewriter<Map<Slot, Alias>> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/RewriteAliasToChildExpr.java:30]: use ExpressionRewriter pls😁

**Review by morrySnow — COMMENTED** [review]
need rebase to newest master code

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/RewriteAliasToChildExpr.java (hunk 15 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/RewriteAliasToChildExpr.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java (hunk 14 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+public class SwapFilterAndProject extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalProject()).thenApply(ctx -> {
+            LogicalFilter<LogicalProject<GroupPlan>> filter = ctx.root;
+            LogicalProject<GroupPlan> project = filter.child();
+            List<NamedExpression> namedExpressionList = project.getProjects();
+            Map<Expression, Expression> slotToAlias = new HashMap<>();
+            namedExpressionList.stream().filter(Alias.class::isInstance).forEach(s -> {
+                slotToAlias.put(s.toSlot(), s);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:48]:
```suggestion
                slotToAlias.put(s.toSlot(), s.child());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            });
+            ExpressionReplacer expressionReplacer = new ExpressionReplacer();
+            Expression rewrittenPredicate = expressionReplacer.visit(filter.getPredicates(), slotToAlias);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:51]:
```suggestion
            Expression rewrittenPredicate = ExpressionReplacer.INSTANCE.visit(filter.getPredicates(), slotToAlias);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java [hunk 18]
```diff
@@ -44,11 +44,9 @@ public class LogicalOlapScanToPhysicalOlapScan extends OneImplementationRuleFact
...
         return logicalOlapScan().then(olapScan ->
             new PhysicalOlapScan(
-                olapScan.getTable(),
-                olapScan.getQualifier(),
-                olapScan.getSelectedIndexId(),
-                olapScan.getSelectedTabletId(),
-                olapScan.getSelectedPartitionId(),
+                olapScan.getTable(), olapScan.getQualifier(), olapScan.getSelectedIndexId(),
+                    olapScan.getSelectedTabletId(),
+                    olapScan.getSelectedPartitionIds(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:49]:
```suggestion
                olapScan.getTable(),
                olapScan.getQualifier(),
                olapScan.getSelectedIndexId(),
                olapScan.getSelectedTabletId(),
                olapScan.getSelectedPartitionIds(),
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 19]
```diff
@@ -42,14 +42,24 @@ public class LogicalOlapScan extends LogicalRelation {
...
 
     public LogicalOlapScan(OlapTable table) {
         this(table, ImmutableList.of());
     }
 
     public LogicalOlapScan(OlapTable table, List<String> qualifier) {
-        this(table, qualifier, Optional.empty(), Optional.empty());
+        this(table, qualifier,
+                Optional.empty(), Optional.empty(),
+                table.getPartitionIds(), false);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:54]:
```suggestion
        this(table, qualifier, Optional.empty(), Optional.empty(), table.getPartitionIds(), false);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 20]
```diff
@@ -42,14 +42,24 @@ public class LogicalOlapScan extends LogicalRelation {
...
+    }
+
+    public LogicalOlapScan(Table table) {
+        this(table, ImmutableList.of());
+    }
+
+    public LogicalOlapScan(Table table, List<String> qualifier) {
+        this(table, qualifier, Optional.empty(), Optional.empty(), ((OlapTable) table).getPartitionIds(), false);
     }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:63]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 21]
```diff
@@ -42,14 +42,24 @@ public class LogicalOlapScan extends LogicalRelation {
...
     public LogicalOlapScan(OlapTable table, List<String> qualifier) {
-        this(table, qualifier, Optional.empty(), Optional.empty());
+        this(table, qualifier,
+                Optional.empty(), Optional.empty(),
+                table.getPartitionIds(), false);
+    }
+
+    public LogicalOlapScan(Table table) {
+        this(table, ImmutableList.of());
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:59]: remove this constructor pls

File: fe/fe-core/src/main/java/org/apache/doris/planner/OlapScanNode.java [hunk 22]
```diff
@@ -231,7 +231,6 @@ public void setTupleIds(ArrayList<TupleId> tupleIds) {
         this.tupleIds = tupleIds;
     }
 
-    // only used for UT
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/OlapScanNode.java:234]: add a comment
```
only used for UT and Nereids
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java (hunk 16 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:48]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java (hunk 17 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:51]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java (hunk 18 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:49]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java (hunk 20 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:63]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java (hunk 19 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:54]: done

File: fe/fe-core/src/main/java/org/apache/doris/planner/OlapScanNode.java (hunk 22 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/OlapScanNode.java:234]: done

**Review by morrySnow — APPROVED** [review]
approved, but some names need to fix, pls fix it in column prune PR.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Rewrite filter -> project to project -> filter.
+ */
+public class SwapFilterAndProject extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalProject()).thenApply(ctx -> {
+            LogicalFilter<LogicalProject<GroupPlan>> filter = ctx.root;
+            LogicalProject<GroupPlan> project = filter.child();
+            List<NamedExpression> namedExpressionList = project.getProjects();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SwapFilterAndProject.java:45]:
```suggestion
            List<NamedExpression> projects = project.getProjects();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 24]
```diff
@@ -58,27 +64,17 @@ public LogicalOlapScan(OlapTable table, List<String> qualifier) {
      * @param table Doris table
      * @param qualifier table name qualifier
      */
-    public LogicalOlapScan(Table table, List<String> qualifier,
-            Optional<GroupExpression> groupExpression, Optional<LogicalProperties> logicalProperties) {
-        super(PlanType.LOGICAL_OLAP_SCAN, table, qualifier, groupExpression, logicalProperties);
+    public LogicalOlapScan(Table table, List<String> qualifier, Optional<GroupExpression> groupExpression,
+            Optional<LogicalProperties> logicalProperties, List<Long> selectedPartitionIdList,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:68]:
```suggestion
            Optional<LogicalProperties> logicalProperties, List<Long> selectedPartitionIds,
```

**Pull request merged**